### PR TITLE
parameterize pipelineTask `displayName` 

### LINKED ISSUE #7200: introduce an option to parameterize `displayName`
`displayName` was introduced as part of https://github.com/tektoncd/community/blob/main/teps/0047-pipeline-task-display-name.md. 

`displayName` helps describing and identifying `pipelineTasks` using human readable names in addition to machine friendly. `displayName` is limited to static values. It will help to specify these values dynamically through `params` in such way that the users can specify their own desired names through `pipelineRun` in a shared env.

For example:

```
displayName: "comments=$(params.comments)"
```

/kind feature

### PR DESCRIPTION
**Pull request by pritidesai**
# Changes

Allow specifying parameters, task results, or context variables in a `displayName`.

```
spec:
params:
  - name: git-repo
  tasks:
    - name: scan
      displayName: "Code Scan for $(params.git-repo)"
      taskRef:
        name: sonar-scan
```

Specifying task results in a `displayName` does not introduce a resource dependency. Pipeline authors have to specify resource dependency using `runAfter` or task `params`.

/kind feature

Fixes #7200 

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
displayName can now accept params, task result references, or context variables.
```

### CODE REVIEW

**Review by jerop — COMMENTED** [review]
this is only applying the replacements on the pipeline task display name, do we want to do the same for pipeline display name? this can be done in future work, just want to understand if it's intentional to not do it for pipelines

please update the docs for displayname in pipelinetasks: https://github.com/tektoncd/pipeline/blob/97184c36a69ef3be0b1e42bd73b57e00f27fcb79/docs/pipelines.md?plain=1#L93

**Review by Yongxuanzhang — APPROVED** [review]

File: examples/v1/pipelineruns/pipeline-with-displayname.yaml [hunk 1]
```diff
@@ -42,7 +42,7 @@ spec:
       default: "1"
   tasks:
     - name: sum-two-numbers
-      displayName: Calculate the first two numbers
+      displayName: "Calculate the first two numbers $(params.a)"
```
[the remark this answers is not in the document]
Comment (pritidesai) [reply examples/v1/pipelineruns/pipeline-with-displayname.yaml]: yes, we can but the resulting string is not normalized/evaluated and will be displayed as:

```
"Calculate the first two numbers 1 + 2"
```

**Comment by pritidesai** [comment]
> this is only applying the replacements on the pipeline task display name, do we want to do the same for pipeline display name?

Yes, this PR is only replacing `displayName` for `tasks[].displayName` or `finally[].displayName`.

We could definitely extend this support to pipeline and task names if there is a use case.
 
> please update the docs for displayname in pipelinetasks:
> 
> https://github.com/tektoncd/pipeline/blob/97184c36a69ef3be0b1e42bd73b57e00f27fcb79/docs/pipelines.md?plain=1#L93

I have updated the documentation, PTAL 🙏

**Review by jerop — COMMENTED** [review]
/lgtm

**Comment by AverageMarcus** [comment]
Does this also work within Matrixes? If so can we close off https://github.com/tektoncd/pipeline/issues/7082 ?

**Comment by pritidesai** [comment]
> Does this also work within Matrixes? If so can we close off #7082 ?

This works with matrix but not as expected by #7082. 

The way we have matrix designed and implemented by creating a list of taskRuns as part of a single pipelineTask, doesn't help to identify different instances of matrix using `pipelineTask.displayName`. `pipelineTask.displayName` stays common across all instances of `matrix`. I have been playing around and trying to figure out the best way to handle `matrix` use case without introducing a new API specifications.

For example:

```
  pipelineSpec:
    tasks:
      - name: platforms-and-browsers
        displayName: "Execution of a task with $(params.platform) and $(params.browser)"
        matrix:
          params:
            - name: platform
              value:
                - linux
                - mac
                - windows
            - name: browser
              value:
                - chrome
                - safari
                - firefox
        taskRef:
          name: platform-browsers
```

The `platforms-and-browsers` pipelineTask is a single task with 9 different taskRuns. All 9 taskRuns rely on a common `displayName` instead of having its own `displayName` for  #7082 to work.

I think we might have to go with your proposed design of having `displayName` under `matrix` and have the dashboard read  in displayName from the matrix section. @AlanGreene thoughts?

**Comment by AlanGreene** [comment]
Having `displayName` under matrix seems like the best option right now. We'd need to clearly document that `pipelineTask.matrix.displayName` takes precedence if `pipelineTask.displayName` is also set, and maybe produce a warning in this case?

As discussed on Slack, ideally the resolved displayName with param / results references replaced would be made available somewhere for all consuming clients (Dashboard, tkn, OpenShift console, etc.) so they don't have to duplicate the effort of resolving it themselves. This could potentially be in `pipelineRun.status.childReferences` or similar.

**Comment by pritidesai** [comment]
I will take this discussion to the issue #7082 so that we can track the history.

**Pull request merged**